dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/mvc/interceptor/AuthorityInterceptor.java

```diff
@@ -0,0 +1,96 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.mvc.interceptor;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.Nonnull;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import lombok.extern.slf4j.Slf4j;
+import org.hisp.dhis.security.Authorities;
+import org.hisp.dhis.security.HasAuthority;
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Component;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.HandlerInterceptor;
+
+/**
+ * Interceptor which checks that the {@link org.hisp.dhis.user.User} has any one of {@link
+ * org.hisp.dhis.security.Authorities} which are passed in {@link HasAuthority} as an arg.
+ *
+ * <p>Throws {@link AccessDeniedException} if {@link org.hisp.dhis.user.User} does not have any of
+ * the passed-in {@link org.hisp.dhis.security.Authorities}.
+ *
+ * <p>{@link Authorities#ALL} is automatically added to the check, as having this Authority allows
+ * access to all methods by default.
+ */
+@Slf4j
+@Component
+public class AuthorityInterceptor implements HandlerInterceptor {
+
+  @Override
+  public boolean preHandle(
+      @Nonnull HttpServletRequest request,
+      @Nonnull HttpServletResponse response,
+      @Nonnull Object handler) {
+    HandlerMethod handlerMethod = (HandlerMethod) handler;
+
+    if (!handlerMethod.hasMethodAnnotation(HasAuthority.class)) {
+      return true;
+    }
+
+    HasAuthority hasAuthority = handlerMethod.getMethodAnnotation(HasAuthority.class);
+    if (hasAuthority != null) {
+      // include 'ALL' authority in required authorities
+      List<Authorities> requiredAuthorities = new ArrayList<>(List.of(Authorities.ALL));
+      requiredAuthorities.addAll(Arrays.stream(hasAuthority.anyOf()).toList());
+
+      // get user authorities
+      final SecurityContext securityContext = SecurityContextHolder.getContext();
+      Authentication authentication = securityContext.getAuthentication();
+      Collection<? extends GrantedAuthority> userAuthorities = authentication.getAuthorities();
+
+      // check if user has any of the required authorities passed in
+      if (requiredAuthorities.stream()
+          .noneMatch(
+              reqAuth ->
+                  userAuthorities.stream()
+                      .anyMatch(userAuth -> reqAuth.name().equals(userAuth.getAuthority())))) {
+        throw new AccessDeniedException("Access is denied");
```

Comment:
Should we name the list of authorities in the message?

Reply:
Could do, whatever we decide to go with 👍

Reply:
after team meeting we've decided to list missing authorities in these checks. Will update in separate PR after speaking with FE.